### ISSUE TITLE
Handshake Inactivity Timeout

### ISSUE THREAD

**Issue by patientplatypus6** [mysqljs/mysql#2456]
I've got a handshake inactivity timeout error using mysql over docker-compose. I've increased the timeout time and that hasn't seemed to work. I don't think it's an issue with docker-compose file as it was working before - I'm pretty sure no file/network changes occurred before this issue started. Does anyone have any ideas on what to try?

EDIT: The answer supplied here: https://stackoverflow.com/questions/35553432/error-handshake-inactivity-timeout-in-node-js-mysql-module/65721732#65721732 doesn't work for example (not sure why the downvote, it doesn't work...). I've posted the question here: https://stackoverflow.com/questions/65724514/error-handshake-inactivity-timeout-mysql-node-js. Please, any help would be appreciated. I've used your database in my code, and my program won't work unless I can get this solved.

For more information, the below is how I am connecting to the database (the `testconnection` is used because `createPool` does not seem to have a debug option - however I am not getting packet information on `testconnection` either).

```
function connectionFunc(){
  var mysql = require('mysql');
  var testconnection = mysql.createConnection({
    host: 'db',
    port: '3306',
    user: 'root',
    password: process.env.MYSQL_ROOT_PASSWORD,
    database: 'blog',
    debug: true
  })
  var pool  = mysql.createPool({
    connectionLimit : 100,
    host            : 'db',
    port            : '3306',
    //setting timeouts here does not work - I've tried.
    //connectTimeout  : 20000,
    //acquireTimeout  : 60 * 60 * 1000,
    //timeout         : 60 * 60 * 1000,
    user            : 'root',
    password        : process.env.MYSQL_ROOT_PASSWORD,
    database        : 'blog'
  });
  console.log("value of testconnection: ", testconnection);
  testconnection.query("SELECT * from blogs", function(err, rows, fields){
    console.log("value of err: ", err);
    console.log("value of rows: ", rows);
    console.log("value of fields: ", fields);
  })
  console.log("value of pool: ", pool);
  let poolPromise = new Promise(resolve=>{
    pool.query("SELECT * FROM blogs", function(err,rows,fields){
      console.log("value or err: ", err);
      console.log('value of rows: ', rows);
      console.log('value of fields: ', fields);
      if (err){
        process.exit();
        resolve(false);
      }else{
        resolve(pool);
      }
    })
  })

  return poolPromise;
}   
    
module.exports = {
  connectionFunc
}  
```

And here is the output: 
```
api      | value of testconnection:  <ref *1> Connection {
api      |   _events: [Object: null prototype] {},
api      |   _eventsCount: 0,
api      |   _maxListeners: undefined,
api      |   config: ConnectionConfig {
api      |     host: 'db',
api      |     port: '3306',
api      |     localAddress: undefined,
api      |     socketPath: undefined,
api      |     user: 'root',
api      |     password: 'NEST',
api      |     database: 'blog',
api      |     connectTimeout: 10000,
api      |     insecureAuth: false,
api      |     supportBigNumbers: false,
api      |     bigNumberStrings: false,
api      |     dateStrings: false,
api      |     debug: true,
api      |     trace: true,
api      |     stringifyObjects: false,
api      |     timezone: 'local',
api      |     flags: '',
api      |     queryFormat: undefined,
api      |     pool: undefined,
api      |     ssl: false,
api      |     localInfile: true,
api      |     multipleStatements: false,
api      |     typeCast: true,
api      |     maxPacketSize: 0,
api      |     charsetNumber: 33,
api      |     clientFlags: 455631
api      |   },
api      |   _socket: undefined,
api      |   _protocol: Protocol {
api      |     _events: [Object: null prototype] {},
api      |     _eventsCount: 0,
api      |     _maxListeners: undefined,
api      |     readable: true,
api      |     writable: true,
api      |     _config: ConnectionConfig {
api      |       host: 'db',
api      |       port: '3306',
api      |       localAddress: undefined,
api      |       socketPath: undefined,
api      |       user: 'root',
api      |       password: 'NEST',
api      |       database: 'blog',
api      |       connectTimeout: 10000,
api      |       insecureAuth: false,
api      |       supportBigNumbers: false,
api      |       bigNumberStrings: false,
api      |       dateStrings: false,
api      |       debug: true,
api      |       trace: true,
api      |       stringifyObjects: false,
api      |       timezone: 'local',
api      |       flags: '',
api      |       queryFormat: undefined,
api      |       pool: undefined,
api      |       ssl: false,
api      |       localInfile: true,
api      |       multipleStatements: false,
api      |       typeCast: true,
api      |       maxPacketSize: 0,
api      |       charsetNumber: 33,
api      |       clientFlags: 455631
api      |     },
api      |     _connection: [Circular *1],
api      |     _callback: null,
api      |     _fatalError: null,
api      |     _quitSequence: null,
api      |     _handshake: false,
api      |     _handshaked: false,
api      |     _ended: false,
api      |     _destroyed: false,
api      |     _queue: [],
api      |     _handshakeInitializationPacket: null,
api      |     _parser: Parser {
api      |       _supportBigNumbers: false,
api      |       _buffer: <Buffer >,
api      |       _nextBuffers: [BufferList],
api      |       _longPacketBuffers: [BufferList],
api      |       _offset: 0,
api      |       _packetEnd: null,
api      |       _packetHeader: null,
api      |       _packetOffset: null,
api      |       _onError: [Function: bound handleParserError],
api      |       _onPacket: [Function: bound ],
api      |       _nextPacketNumber: 0,
api      |       _encoding: 'utf-8',
api      |       _paused: false
api      |     },
api      |     [Symbol(kCapture)]: false
api      |   },
api      |   _connectCalled: false,
api      |   state: 'disconnected',
api      |   threadId: null,
api      |   [Symbol(kCapture)]: false
api      | }
api      | value of pool:  <ref *1> Pool {
api      |   _events: [Object: null prototype] {},
api      |   _eventsCount: 0,
api      |   _maxListeners: undefined,
api      |   config: PoolConfig {
api      |     acquireTimeout: 10000,
api      |     connectionConfig: ConnectionConfig {
api      |       host: 'db',
api      |       port: '3306',
api      |       localAddress: undefined,
api      |       socketPath: undefined,
api      |       user: 'root',
api      |       password: 'NEST',
api      |       database: 'blog',
api      |       connectTimeout: 10000,
api      |       insecureAuth: false,
api      |       supportBigNumbers: false,
api      |       bigNumberStrings: false,
api      |       dateStrings: false,
api      |       debug: undefined,
api      |       trace: true,
api      |       stringifyObjects: false,
api      |       timezone: 'local',
api      |       flags: '',
api      |       queryFormat: undefined,
api      |       pool: [Circular *1],
api      |       ssl: false,
api      |       localInfile: true,
api      |       multipleStatements: false,
api      |       typeCast: true,
api      |       maxPacketSize: 0,
api      |       charsetNumber: 33,
api      |       clientFlags: 455631
api      |     },
api      |     waitForConnections: true,
api      |     connectionLimit: 100,
api      |     queueLimit: 0
api      |   },
api      |   _acquiringConnections: [],
api      |   _allConnections: [],
api      |   _freeConnections: [],
api      |   _connectionQueue: [],
api      |   _closed: false,
api      |   [Symbol(kCapture)]: false
api      | }
api      | here are the environment variables passed in from docker-compose: 
api      | MYSQL_ROOT_PASSWORD:  <secret>
api      | MYSQL_PASSWORD:  <secret>
```

The error specifically - 

```
api      | value or err:  Error: Handshake inactivity timeout
api      |     at Handshake.<anonymous> (/app/node_modules/mysql/lib/protocol/Protocol.js:160:17)
api      |     at Handshake.emit (node:events:376:20)
api      |     at Handshake._onTimeout (/app/node_modules/mysql/lib/protocol/sequences/Sequence.js:124:8)
api      |     at Timer._onTimeout (/app/node_modules/mysql/lib/protocol/Timer.js:32:23)
api      |     at listOnTimeout (node:internal/timers:556:17)
api      |     at processTimers (node:internal/timers:499:7)
api      |     --------------------
api      |     at Protocol._enqueue (/app/node_modules/mysql/lib/protocol/Protocol.js:144:48)
api      |     at Protocol.handshake (/app/node_modules/mysql/lib/protocol/Protocol.js:51:23)
api      |     at PoolConnection.connect (/app/node_modules/mysql/lib/Connection.js:116:18)
api      |     at Pool.getConnection (/app/node_modules/mysql/lib/Pool.js:48:16)
api      |     at Pool.query (/app/node_modules/mysql/lib/Pool.js:202:8)
api      |     at /app/db/connect.js:32:10
api      |     at new Promise (<anonymous>)
api      |     at Object.connectionFunc (/app/db/connect.js:31:21)
api      |     at Object.<anonymous> (/app/routes/index.js:17:12)
api      |     at Module._compile (node:internal/modules/cjs/loader:1108:14) {
api      |   code: 'PROTOCOL_SEQUENCE_TIMEOUT',
api      |   fatal: true,
api      |   timeout: 10000
api      | }
```

The system information I have is the following: 

inxi
CPU Single core Intel Xeon E5-2650 v4 (-MCP-) speed 2199 MHz (max) Kernel 4.15.0-130-generic x86_64 Up 13:40 Mem 792.2/985.2MB HDD 26.8GB(36.1% used) Procs 136 Client Shell inxi 2.3.56 

running ubuntu 18.04

**Comment by patientplatypus6** [comment]
Ah! As it happens, I had a default network in my docker-compose. So, the db apparently doesn't communicate over the default network. Huh. When I take that out it works! But the default network was being used to circumvent another unrelated issue, so I'll pursue that if necessary. Thanks for everybody that looked at my stackoverflow - I'll update that as well. Cheers!

**Comment by patientplatypus6** [comment]
Still breaking - I've updated the stackoverflow question: https://stackoverflow.com/questions/65724514/error-handshake-inactivity-timeout-mysql-node-js.

**Comment by patientplatypus6** [comment]
Cross posted a docker-compose error on their github as well. Thanks!

**Comment by patientplatypus6** [comment]
Uninstalled and reinstalled docker and that hasn't worked.

**Comment by jgomes94** [comment]
@patientplatypus6  do you have any update on that? I believe I'm also facing this issue

**Comment by Zikoel** [comment]
Just for brainstorming: do you have done tests with other libraries or tools? The fact that the error is raised from this library tell us that this library is the driver, not the source of the problem! First of all a good test to do is verify if with other tools in the same situation the error persist (maybe a container with mysql-cli)! An handshake error can be an error at a deep point of this library or a network problem! Here you are playing with docker networks so there is a chance that the problem is there.

**Comment by kastman** [comment]
I'm also facing this edge case; I've done testing inside the container, and can both ping the db container and connect with mysql client, though node hangs. I see you're using pool, but I can't connect even with a single connection. Adding `debug: true` to the options doesn't help, and using the IP of the db on the bridged network doesn't work either. It's really weird. I looked for proxy options, or anything else that's similar; the only thing I see is that I'm also using the `--default-authentication-plugin=mysql_native_password` auth type?

Any guidance on what kind of traces we can collect, to hunt this down?

**Comment by ds-arb** [comment]
@kastman were you able to solve this issue?

**Comment by patientplatypus6** [comment]
I'm currently working on a different problem, but if anyone needs me to try and boot the example again and take a look please just (private) message me (or email at patientplatypus6@gmail.com). I'm getting updates on this because I'm tagged to the thread. I can (probably) at least make a repo showing the issue. For the moment, I'm closing this as there doesn't seem to be all that much activity. Feel free to open another issue. Thanks!

**Comment by kastman** [comment]
Sorry @ds-arb - I never made progress and went back to running some services locally outside of docker containers. I'm with @patientplatypus6 - would be interested in a solution but not able to work on it now.

**Comment by kastman** [comment]
@ds-arb and @patientplatypus6 -- I was actually able to look at this further, and realized that there was a real authentication error which was being hidden by the timeout, I suspect possibly induced by express or another error handler. My authentication problems came from the fact that even though I was using `command: --default-authentication-plugin=mysql_native_password` in my dockerfile to boot the server with an older password scheme, I think the initial setup performed as part of the mysql image was prior to this and not respecting this command arguments (e.g. the environment variables for `MYSQL_USER` and `MYSQL_PASSWORD` were being set with the standard mysql 8 authentication scheme. The error that was being hidden for me was `ER_NOT_SUPPORTED_AUTH_MODE: Client does not support authentication protocol requested by server; consider upgrading MySQL client`, so changing my client from mysql to mysql2 which supports the standard auth was all I needed to do. 

I figured this out primarily by running a small separate POC that only handled the mysql parts, and didn't do any of the error handling or express serving; then the error popped up immediately. Hope that helps, and good luck,
